### PR TITLE
Add flat_map in stdlib

### PR DESCRIPTION
**Pull request by xvw**
The `flat_map` function is already available in the `Seq` module (which also includes a `concat_map` function as an alias) and `Result` and `Option` introduce `bind` where `flat_map` is a flipped alias.

The main reason is that when I'm prototyping a small project with few dependencies (so no infix operators or binding operators), the order of the parameters makes function piping laborious. For example : 

```ocaml
let _ = x |> Result.map f |> (fun r -> Result.bind r g) (* or Fun.flip Result.bind *)
(* becomes *)
let _ = x |> Result.map f |> Result.flat_map g
```

I think the name `flat_map` is fairly common (used in particular by the Scala community and by the `Seq` module). And I have a lot of cases where I end up rebuilding `let*` or `>>=` because I find that `bind` makes pipelines hard to read.

### CODE REVIEW

**Comment by dbuenzli** [comment]
I'm not very convinced by your example. 

For me this use of the pipe operator is rather bad style and longwinded (like most examples of the pipe operators I see in the wild). I find either of these more readable:

```ocaml
let _ = M.join (M.map (Fun.compose g f) x)
let _ = M.join @@ M.map (Fun.compose g f) x
```
 
And in general `M.join (M.map f v)` clearer than introducing more names to learn.

**Comment by dbuenzli** [comment]
> I'm prototyping a small project with few dependencies (so no infix operators or binding operators)

Also no need to introduce dependencies for that a: 
```
let ( let* ) = Result.bind
```
at the top of your source will do.

**Comment by xvw** [comment]
> I'm not very convinced by your example.
> 
> For me this use of the pipe operator is rather bad style and longwinded (like most examples of the pipe operators I see in the wild). I find either of these more readable:
> 
> ```ocaml
> let _ = M.join (M.map (Fun.compose g f) x)
> let _ = M.join @@ M.map (Fun.compose g f) x
> ```
> 
> And in general `M.join (M.map f v)` clearer than introducing more names to learn.

To be honest, I disagree (but it is just my opinion), for me having a pipeline of function is readable and straigthfoward. (and `Fun.compose` is not part of the stdlib)

> > I'm prototyping a small project with few dependencies (so no infix operators or binding operators)
> 
> Also no need to introduce dependencies for that a:
> 
> ```
> let ( let* ) = Result.bind
> ```
> 
> at the top of your source will do.

Yes, and I need also to have `let (let+) x f = Result.map` which can lead to add a new module to have my own standard library (which can be not an issue, I agree). 

My suggestion if motivated by some examples and yes, there is always quick solution. But here, `flat_map` is already present in the stdlib and the name is accepted by other popular functional programming languages.

**Comment by dbuenzli** [comment]
> To be honest, I disagree (but it is just my opinion), for me having a pipeline of function is readable and straigthfoward

Any code of the form `let y = x |> f` just makes me cringe. But I understand it's natural for people who are used to program imperatively – and hence likely why it became so popular.

To me except in a few cases it's mostly not a good idea. When you read `let` bindings it is much nicer to get a quick idea to what kind of value the variable is bound to without having to look for the end of the expression. This makes understanding a binding a form of progressive disclosure.

Basically use of the pipe operator mostly makes it's backward for the reader to understand code, it lacks abstraction. Also note that in shell programming the named result is mostly at the end `p0 | p1 > bla.txt` so it works there.

> But here, `flat_map` is already present in the stdlib

It's rather [an anomaly](https://github.com/ocaml/ocaml/pull/8760#issuecomment-515059721).

**Comment by xvw** [comment]
I appreciate your feedback, and I confess to having few counter-examples beyond my own personal experience, where sometimes chaining by `|>` avoids the introduction of intermediate names, which would also be possible by using dedicated infix operators that aren't part of the standard library.

So I propose to collect different opinions. If no one is happy with the PR, I'll close it down.

**Comment by nojb** [comment]
> So I propose to collect different opinions. If no one is happy with the PR, I'll close it down.

I appreciate the civil tone of the discussion. To be clear, additions to the standard library need to clear a relatively high bar of consensus because once something is in it, we need to live with it for a long time. So even additions that may be reasonable in isolation or in specific contexts may still have trouble being accepted for inclusion in the standard library. The formal requirement is the approval by two core developers of any contribution that touches the standard library API.

My own opinion about `|>` (which somehow motivates the present PR) is that it encourages an inefficient style in the hands of beginners (eg doing map after map after map instead of a single map of a composed function) and that it is generally inferior to `let` as a sequencing device, but other programmers I respect have different opinions about this operator, so my conclusion is that usability of such operators is very much a subjective matter.

I don't have a strong opinion on the PR itself either way.

**Comment by Octachron** [comment]
I have the feeling that closing the PRs might be slightly premature?
Two days is short reaction time to answer for many people. I would propose to wait for one week?

I am personally tepidly favourable to the proposition: I think that left-to-right application order is a good fit for a sequence of transformation when it is the set of transformations that should be highlighted rather than the end result or some nameless intermediary steps.

In this context, `flat_map` is slightly more usable than `bind`, and the name exists and might more familiar than `bind`  for non-functional oriented programmers.

At the same time, I also agree that it is problematic to have too many synonymous functions in the standard library.

**Comment by xvw** [comment]
I found the arguments against acceptable, even if, as I did give an example with `|>`, it's not really the heart of the PR. 
Even `x |> map |> bind` is not efficient, if we focus on `flat_map` and not `|>` I can see a number of good points:

**pros**:
- coherent with `map` 
- makes partial application more obvious 

**cons**:
- name duplication
- enforce a non-efficient way of programming (using `|>`).

I find it hard to judge the pros and cons. I have memories of having regrested the order of `bind` arguments (as a function and not as an operator), but there are trade-offs that are relatively unobtrusive. so /shrug :)
